### ISSUE TITLE
P0533R9 `constexpr` For `<cmath>` And `<cstdlib>`

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#2530]
[P0533R9](https://wg21.link/P0533R9) `constexpr` For `<cmath>` And `<cstdlib>`

~~LWG-3834 Missing `constexpr` for `std::intmax_t` math functions in `<cinttypes>`~~ (tracked by #6414 instead)

Feature-test macro:
```cpp
#define __cpp_lib_constexpr_cmath 202202L
```

**Comment by CaseyCarter** [comment]
We need to investigate this sooner rather than later to determine exactly how we're going to make UCRT functions `constexpr`. I suspect we'll have to teach the compiler(s) to recognize e.g. `::ldexp` and `std::ldexp`, at least when they appear in constant expressions, and constant-fold them. This is going to involve a lot of joint work and communication with our compiler teams.

**Comment by AlexGuteniev** [comment]
I think it is implementable without compiler support. You have `constexpr bit_cast`, you know the representation of floats, you do the math.

Compiler support looks better though. Be sure to have it in clang-cl either.

**Comment by frederick-vs-ja** [comment]
I don't think compiler support is essentially needed (but may be nice to have). We can ask UCRT not to declare these functions, but to provide some equivalent functions with different names (e.g. `__cstd_fabs`, which may be suitable to be called at runtime) in C++23 mode.

**Comment by fsb4000** [comment]
a few initial thoughts from LLVM discord: https://discord.com/channels/636084430946959380/636732781086638081/942850475752062989

@ldionne wrote:

>I think we'll definitely need some level of compiler support, otherwise we'll need to basically reimplement those functions inside libc++ itself like so:
```c++
constexpr long abs(long n) {
  if (is_constant_evaluated()) {
    // compile-time friendly implementation
  } else {
    return ::labs(n);
  }
}
```

>abs() is a bit too easy, but the same goes for other functions
>
>The main problem I see with this is that we currently use the using_if_exists attribute on all our namespace std { using ::abs; } declarations, and there's no way to do this if we have to define a function
>
>We would need to extend using_if_exists to that it can be applied to other types of declarations, like function declarations, and that the declaration it checks for existence isn't the same as the one being declared. So we'd have to do something like:
```c++
namespace std {
  __attribute__((using_if_exists(::abs))) constexpr long abs(long n) { ... }
}
```

**Comment by StephanTLavavej** [comment]
I've contacted the compiler team about getting support for this.

**Comment by StephanTLavavej** [comment]
We should ask the compiler team to consider implementing #3789 at the same time to save work.

**Comment by blackninja9939** [comment]
The attribute [[msvc::constexpr]] allows "extended constexpr" in C++20 such as calling placement new from within a constexpr function for std::construct_at, could that attribute be extended to allow using these blessed functions in a constant expression?

Seems like the least painful approach in terms of existing blessed constexpr and without needing to reimplement every function manually again

**Comment by frederick-vs-ja** [comment]
> Seems like the least painful approach in terms of existing blessed constexpr and without needing to reimplement every function manually again

I think the crux is the UCRT headers (mainly `<math.h>` in the case of this issue) - they seem somehow "freezed", so it's unclear whether we can even add attributes to declarations in them. As a result, compiler frontends seemingly need work around this when such a function calls is encountered in constant evaluation.

**Comment by frederick-vs-ja** [comment]
> We need to investigate this sooner rather than later to determine exactly how we're going to make UCRT functions `constexpr`. I suspect we'll have to teach the compiler(s) to recognize e.g. `::ldexp` and `std::ldexp`, at least when they appear in constant expressions, and constant-fold them. This is going to involve a lot of joint work and communication with our compiler teams.

Ah, it seems that no progress seem was made since then. If MS had allowed us to touch related UCRT headers or reorganize them like `<stdatomic.h>`, we would probably have implemented the whole library workaround before 2024.
(Also, the alternatename trick does't seem unnecessary. Probably another internal namespace suffices.)

**Comment by StephanTLavavej** [comment]
We're having a meeting on 2025-05-07 with the compiler and UCRT teams to figure out the plan here.

**Comment by YexuanXiao** [comment]
> We're having a meeting on 2025-05-07 with the compiler and UCRT teams to figure out the plan here.

Is there any news to share?

**Comment by StephanTLavavej** [comment]
Not yet, the compiler team is still looking into this before deciding on a course of action. A major concern is how to address accuracy issues (i.e. should the compiler prioritize mathematically-correct results, or UCRT bug-compatibility).

**Comment by h-vetinari** [comment]
> (i.e. should the compiler prioritize mathematically-correct results, or UCRT bug-compatibility).

For the sanity of our future selves, please advocate for prioritizing "correct results". Old UCRT versions will fade away over time, but intentionally buggy core math functions will keep biting people ~forever

**Comment by StephanTLavavej** [comment]
That's what I'm advocating for whenever the compiler team asks me.

**Comment by clin1234** [comment]
Any progress from the compiler team in Q1 2026?

**Comment by StephanTLavavej** [comment]
Yes, I count roughly 33 compiler PRs with a cumulative `git diff --shortstat` of 7581 insertions and 2673 deletions. There's nothing visible to users (or library devs) yet, but they're making tons of progress.

**Comment by clin1234** [comment]
> Yes, I count roughly 33 compiler PRs with a cumulative `git diff --shortstat` of 7581 insertions and 2673 deletions. There's nothing visible to users (or library devs) yet, but they're making tons of progress.

How about the beginning of middle of Q2?

**Comment by StephanTLavavej** [comment]
While no significant commits have landed in MSVC `main` for this since my last comment, the FE team has been making progress with testing the implementation against our "Real World Code" (RWC) test suite of open-source projects. This found some projects that were being bad kitties, jumping up on the table where they're not allowed and declaring math functions by hand (which will now lack `constexpr`, leading to errors).

I am busy with non-STL work this month, but responding to whatever the FE team needs from the STL for `constexpr <cmath>` will be my second-highest priority.
